Default behavior for recordedit

jrchudy opened the pull request:
When a user navigates to recordedit without a catalog, schema, or table defined they are given an error dialog and then redirected to the landing page. This is meant to address issue #442 as the default behavior when catalog, schema, or table is not present in the URL.

This can be tested on my [synapse dev machine](https://synapse-dev.isrd.isi.edu/~jchudy/chaise/data-entry/#200/Zebraish:Subject). That URL points to catalog `200` which does not exist. The same type of dialog can be seen for the case of an improper schema or table name as well or if all of that information is missing, [for instance](https://synapse-dev.isrd.isi.edu/~jchudy/chaise/data-entry/).

Assigning to @robes and @howdyjessie to review the code changes and merge the pull request.

howdyjessie commented:
@jrchudy If I'm logged in, the error dialog appears and clicking OK takes me back to the landing page. But if I'm logged out, the error dialog doesn't show up, and then I see an error about `UriUtils` in the console:

![screen shot 2016-06-22 at 12 38 43 pm](https://cloud.githubusercontent.com/assets/4585431/16280827/563e5ff2-3876-11e6-84f3-a7e9b03e189e.png)

jrchudy commented:
This is a side effect of the refactoring where we have been re-doing the logic for unauthorized. I believe we talked about dealing with 401 by having ermrestJS try to log the user in, but I could be wrong.

jrchudy commented:
I updated the code to include the Chaise catchAll function that also handles the login flow if the user is not logged in and is required to. They are redirected to the login page where they can choose which login provider they want.

This resolves the issue that Jessie noted when looking at the code. This can be tested [here](https://synapse-dev.isrd.isi.edu/~jchudy/chaise/recordedit/#2/Zebrafish:Subject) by first making sure you are logged out before navigating to this page.

robes commented on Makefile:
```diff
@@ -220,6 +220,7 @@ RE_SHARED_JS_DEPS=$(JS)/vendor/jquery-latest.min.js \
 	$(COMMON)/utils.js \
 	$(COMMON)/validators.js \
 	$(COMMON)/navbar.js \
+	$(COMMON)/templates/errorDialog.controller.js \
```
seems odd to see a controller.js in a 'templates' subdirectory?

robes commented on recordedit/recordEdit.app.js:
```diff
@@ -201,23 +196,24 @@
                 // } else if (exception instanceof Errors.SchemaNotFoundError) {
```
probably don't need to preserve the commented code

robes commented:
This all looks fine to me and I tried both of the links (s/data-entry/recordedit). I made a couple minor comments in the code. This looks fine to merge to me.

howdyjessie commented on common/errors.js:
```diff
@@ -4,7 +4,42 @@
+                backdrop: 'static',
+                keyboard: false,
+                resolve: {
+                    params: function() {
```
Is it necessary to return `params` via this function instead of just
```
resolve: {
    params: params
}
```
?

howdyjessie commented on common/templates/errorDialog.controller.js:
```diff
@@ -0,0 +1,15 @@
+
+    angular.module('chaise.errors')
+
+    .controller('ErrorDialogController', ['$uibModalInstance', 'params', function ConfirmDeleteController($uibModalInstance, params) {
```
For consistency, it might be better to change `ConfirmDeleteController` to `ErrorDialogController`.

howdyjessie commented on common/errors.js:
```diff
@@ -4,7 +4,42 @@
+            });
+
+            modalInstance.result.then(function () {
+                $window.location = chaiseConfig.dataBrowser ? chaiseConfig.dataBrowser : $window.location.origin;
```
If we're truly redirecting (are we? cc @robes?) and not just going to another page, then it should be `$window.location.replace(url)`. [window.location vs window.location.replace()](http://stackoverflow.com/questions/1865837/whats-the-difference-between-window-location-and-window-location-replace)

jrchudy commented:
Assigned back to @howdyjessie to make sure changes look good. I did a small refactor for viewer by removing `confirmDelete.controller.js` and instead using `modalDialog.controller.js` in the common folder. Also moved `confirmDelete.html` template to the `common/templates` folder. 

Refactored changes for [recordedit](https://synapse-dev.isrd.isi.edu/~jchudy/chaise/recordedit/#2/Zebrafish:Subject). Try changing the url parameters or removing them completely. As well as being logged out when trying to navigate to the page.

Refactored changes for [viewer](https://dev.rebuildingakidney.org/~jchudy/chaise/viewer/#1/rbk:image/id=7). Make a comment and try to delete it. As well as Make  new annotation and try to delete that with and without a comment. All of this should still work.

howdyjessie commented:
Everything looks good to me; I'm merging this PR.
